### PR TITLE
[JIT] Hacky: Staged symbolics for RNN nodes

### PR DESCRIPTION
**Pull request by jamesr66a**
Stacked on top of https://github.com/pytorch/pytorch/pull/6256

Similarly to that PR, this PR makes it so that RNN nodes (e.g. LSTM) lower into a PythonOp with an associated `symbolic` attribute containing the ONNX symbolic for later export.

This PR brings up a few issues: 

* The RNN functions take in and return (nested) tuples and lists of Tensors. In order to actually run the Python implementation given the flattened argument list presented by the JIT, I have to wrap the RNN functions in another function that rematerializes those Python data structures. This is hacky but it works
* Eventually we're probably going to want to convert the RNN modules into `ScriptModule`s so that we can optimize them with the rest of the graph.

### CODE REVIEW

**Comment by jamesr66a** [comment]
onnx-fb-universe broken on master. Fix here: https://github.com/onnxbot/onnx-fb-universe/pull/1525/files

File: torch/nn/_functions/rnn.py [hunk 1]
```diff
@@ -309,8 +310,68 @@ def forward(input, *fargs, **fkwargs):
         import torch
         if torch._C._jit_is_tracing(input):
             import torch.onnx.symbolic
+            sym = torch.onnx.symbolic.RNN_symbolic_builder(*args, **kwargs)
+            cell_type = args[0]
+
+            def rnn_trace_override_symbolic(g, input, weights, hiddens, batch_sizes):
```
Comment (apaszke) [inline torch/nn/_functions/rnn.py]: I guess it's ok to have 10 line symbolics in different files, but inlining this into `RNN` is too much. Can this be moved to `torch.onnx` somehow? It basically reimplements the RNN which is not great.

**Comment by jamesr66a** [comment]
I don't know where this `onnx` submodule update is coming from

**Comment by jamesr66a** [comment]
@onnxbot retest this please

**Comment by jamesr66a** [comment]
@onnxbot retest this please

**Comment by jamesr66a** [comment]
@onnxbot retest this please

**Comment by jamesr66a** [comment]
@onnxbot retest this please

**Comment by jamesr66a** [comment]
@zdevito test comment please ignore

**Comment by jamesr66a** [comment]
@onnxbot retest this please

**Comment by jamesr66a** [comment]
sweet jesus the onnx test actually worked

**Review by zdevito — APPROVED** [review]
This looks good to me now. Context: we need this so that traced rnns can still run while debugging in script mode. Eventually the override will itself be a script function as well.

File: torch/onnx/symbolic.py [hunk 2]
```diff
@@ -852,3 +857,62 @@ def symbolic(g, input, all_weights, h0, batch_sizes):
         return prev_output, h_outs
 
     return symbolic
+
+
+def rnn_trace_override_symbolic(cell_type, func, sym, g, input, weights, hiddens, batch_sizes):
```
Comment (apaszke) [inline torch/onnx/symbolic.py]: Some extra comments would be nice. E.g. what `func` is, what's the format of `weights`, `hiddens`, what are their types?

**Review by apaszke — APPROVED** [review]
We can merge this if you want (since it's temporary), but I don't understand what's happening in `rnn_trace_override_symbolic`, because it's quite dense and has no docs.

**Pull request merged**